src/Text/LLVM/AST.hs

```diff
@@ -1647,10 +1647,14 @@ data DISubprogram' lab = DISubprogram
 
 type DISubprogram = DISubprogram' BlockLabel
 
-data DISubrange = DISubrange
-  { disrCount      :: Int64
-  , disrLowerBound :: Int64
-  } deriving (Data, Eq, Generic, Ord, Show, Typeable)
+data DISubrange' lab = DISubrange
+  { disrCount      :: Either Int64 (Maybe (ValMd' lab))
+  , disrLowerBound :: Either Int64 (Maybe (ValMd' lab))
```

Comment:
Using `Either` here is somewhat grotesque. As an alternative, what about always encoding these as `Maybe (ValMd' lab)`? On old versions of LLVM, we can simply use `Just (ValMdValue (Typed (PrimType Integer) (ValInteger ...)))`

Reply:
Based on the comment from the llvm-project here: https://github.com/llvm/llvm-project/blob/bbe8cd13335300958b04db5318c31ff52714f96f/llvm/lib/Bitcode/Reader/MetadataLoader.cpp#L1437-L1444
it seems more authentic to the actual representation to use this Either encoding.  While we could potentially push the value down into the `ValMd` like you are suggesting, I think it obfuscates the dual-usage of this field and has possible issues for the pretty-printing side of things:

* It implies that any ValMd encoding could be used
* It creates partiality in the pretty-printing if a ValMd representation is chosen that cannot be represented as an inline signed integer (Integer -> ValMd is injective, but ValMd -> Integer is not).

As noted in my response to your previous comment above, the current pretty-printing will need to be enhanced to handle the (presumably older) LLVM versions that use the signed integer encoding rather than the ValMd... once we know what those versions are.  Right now there are about 6 PR's needing to land to get to that base operational point and I'd like to avoid introducing additional complexity until we get there.

Reply:
> * It implies that any ValMd encoding could be used

I don't think so? Even with the `Either Int64 (Maybe (ValMd' lab))` encoding, there are only a few possible `ValMd'` values that can appear here in practice. For example, the commit which [made `discrLowerBound` an arbitrary metadata node](https://github.com/llvm/llvm-project/commit/d20bf5a7258d4b6a7f017a81b125275dac1aa166) notes that it can only be a signed integer, a `DIVariable`, or a `DIExpression`. Either way, it is possible to have unrepresentable states. 

> * It creates partiality in the pretty-printing if a ValMd representation is chosen that cannot be represented as an inline signed integer (Integer -> ValMd is injective, but ValMd -> Integer is not).

I don't understand what you mean here: pretty-printing a `ValMd` is a total function. It's possible to have unrepresentable states that could be pretty-printed in ways that would result in invalid `discrLowerBound` values, but as noted above, that would happen regardless of how we encode these fields. I think it's better to just accept this, document what forms of `ValMd'` values are allowed here, and pick a simpler encoding (`Maybe (ValMd' lab)`).